src/module.c

```diff
@@ -7418,6 +7443,195 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+ * Multiple callbacks can be registered across different modules. When a Module is unloaded, all the
+ * auth callbacks registered by it are unregistered.
+ * The callbacks are attempted (in the order of most recently registered first) when the AUTH/HELLO
+ * (with AUTH field is provided) commands are called.
+ * The callbacks will be called with a module context along with a username and a password, and are
+ * expected to take one of the following actions:
+ * (1) Authenticate - Use the RM_AuthenticateClient* API and return REDISMODULE_AUTH_HANDLED.
+ * This will immediately end the auth chain as successful and add the OK reply.
+ * (2) Deny Authentication - Return REDISMODULE_AUTH_HANDLED without authenticating or blocking the
+ * client. Optionally, `err` can be set to a custom error message and `err` will be automatically
+ * freed by the server.
+ * This will immediately end the auth chain as unsuccessful and add the ERR reply.
+ * (3) Block a client on authentication - Use the RM_BlockClientOnAuth API and return
+ * REDISMODULE_AUTH_HANDLED. Here, the client will be blocked until the RM_UnblockClient API is used
+ * which will trigger the auth reply callback (provided through the RM_BlockClientOnAuth).
+ * In this reply callback, the Module should authenticate, deny or skip handling authentication.
+ * (4) Skip handling Authentication - Return REDISMODULE_AUTH_NOT_HANDLED without blocking the
+ * client. This will allow the engine to attempt the next custom auth callback.
+ * If none of the callbacks authenticate or deny auth, then password based auth is attempted and
+ * will authenticate or add failure logs and reply to the clients accordingly.
+ *
+ * Note: If a client is disconnected while it was in the middle of blocking custom auth, that
+ * occurrence of the AUTH or HELLO command will not be tracked in the INFO command stats.
+ *
+ * The following is an example of how non-blocking module based authentication can be used:
+ *
+ *      int auth_cb(RedisModuleCtx *ctx, RedisModuleString *username, RedisModuleString *password, RedisModuleString **err) {
+ *           const char* user = RedisModule_StringPtrLen(username, NULL);
+ *           const char* pwd = RedisModule_StringPtrLen(password, NULL);
+ *           if (!strcmp(user,"foo") && !strcmp(pwd,"valid_password")) {
+ *               RedisModule_AuthenticateClientWithACLUser(ctx, "foo", 3, NULL, NULL, NULL);
+ *               return REDISMODULE_AUTH_HANDLED;
+ *           }
+ *           else if (!strcmp(user,"foo") && !strcmp(pwd,"wrong_password")) {
+ *               RedisModuleUser *user = RedisModule_GetModuleUserFromUserName(username);
+ *               if (user) {
+ *                   RedisModuleString *log = RedisModule_CreateString(ctx, "Module Auth", 11);
+ *                   RedisModule_ACLAddLogEntry(ctx, user, log, REDISMODULE_ACL_LOG_AUTH);
+ *                   RedisModule_FreeString(ctx, log);
```

Comment:
This reminds me that we should probably have an API for static lifetime strings so we're not constantly creating them for little stuff like this. Maybe I'll go add it to the backlog.

Reply:
that's a problem since it'll require allocating them on the stack, which would mean that they can't be opaque.

Reply:
I'm not sure we're talking about exactly the same thing. I was also just thinking about a way to expose RedisStrings that have a "shared" refcount, so they could be allocated at init and used repeatedly without memory allocations.